Week 8: OpenMP (#129)
UCL/research-computing-with-cpp: 08openmp/index.md

---
title: "Week 8: Parallel Programming with OpenMP"
---

## Overview

> This week we'll be focusing on speeding up our code by taking advantage of **parallelism**, i.e. splitting up problem and giving different pieces to different processors. In these notes we'll introduce the following concepts:

Reviewer: ```suggestion
This week we'll be focusing on speeding up our code by taking advantage of **parallelism**, i.e. splitting up a problem and giving different pieces to different processors. In these notes we'll introduce the following concepts:
```

- What is parallel programming?
  - Why do we need parallel programming?
  - What are shared- and distributed-memory parallelism?
  - What are weak and strong scaling?
- An introduction to OpenMP
  - Parallelising loops
  - Reductions
  - Data sharing
  - Thread control
  - Schedules
  - Parallelising multi-dimensional loops

## What is parallel programming?

> So far in this course we've been writing fundamentally **sequential** programs, where the code runs from start to finish, each step one after another. Our programs don't try and do multiple operations at once. However, modern **multi-core** CPUs are perfectly capable of performing multiple operations in **parallel**. Our sequential programs are like single queues at a supermarket checkout, capable of processing only one person's items at a time. Open more lanes and we can process a lot more people! If we can split up our problems appropriately and run each piece in parallel, we should be able to speed up our code immensely.

Reviewer: ```suggestion
So far in this course we've been writing fundamentally **sequential** programs, where the code runs from start to finish, each step one after another. Our programs don't try to do multiple operations at once. However, modern **multi-core** CPUs are perfectly capable of performing multiple operations in **parallel**. Our sequential programs are like single queues at a supermarket checkout, capable of processing only one person's items at a time. Open more lanes and we can process a lot more people! If we can split up our problems appropriately and run each piece in parallel, we should be able to speed up our code immensely.
```
This is slightly complicated a bit by the fact that we introduced SIMD registers and operations last week, so there has been a tiny bit of parallelism. You can mention this for completeness but I don't think it's very confusing to gloss over either.

In order to perform truly parallel computation, that is two completely different operations happening at the exact same moment, CPUs had to become **multi-core**, hosting many processing units within one CPU and using them to run many threads in parallel. In consumer PCs this transition started with the dual-core Intel Core2Duo series in the mid-2000s, but supercomputers were already playing with the slightly related idea of **multi-CPU** systems with hundreds of CPUs as early as the 80s! As of 2023, core counts on both server and consumer CPUs can be as high as 64. Typical consumer systems will have only one CPU while servers may have several on one motherboard. Writing parallel programs for multi-core machines is made easy using the focus of this week: **OpenMP**.

> Up until the 90s, supercomputers were dominated by **shared-memory** systems, where all processing units accessed one single memory space. In the late-80s/early-90s it become economically more feasible to increase parallelism and computational power by creating **distributed** systems made up of multiple **nodes**. Each node is a single computer with its own CPU(s), memory space, networking hardware, and possibly other components like hard drives, all networked together in such a way they can communicate information and operate as *one single parallel machine*. I'll start using the term **node** to refer to a single computer as part of a larger, distributed parallel machine. While shared-memory systems still exist[^jq-terabytes], nearly all the world's largest computers are now distributed systems.

Reviewer: ```suggestion
Up until the 90s, supercomputers were dominated by **shared-memory** systems, where all processing units accessed one single memory space. In the late-80s/early-90s it became economically more feasible to increase parallelism and computational power by creating **distributed** systems made up of multiple **nodes**. Each node is a single computer with its own CPU(s), memory space, networking hardware, and possibly other components like hard drives, all networked together in such a way they can communicate information and operate as *one single parallel machine*. I'll start using the term **node** to refer to a single computer as part of a larger, distributed parallel machine. While shared-memory systems still exist[^jq-terabytes], nearly all the world's largest computers are now distributed systems.
```

> Distributed systems tend to have separate memory spaces for each node[^multi-cpu] which introduces a new complexity: how do the nodes share data or communicate when necessary? These **distributed-memory** systems are commonly programmed using a library called the **Message Passing Interface** (MPI) which allows individual nodes to communicate by passing data back and forth. We'll get into MPI properly in the last part of this course. For now, it's useful just to understand the difference between the two basic forms of parallel computing: **shared-memory** parallelism where each thread shares the same memory space and can access any piece of data within it, and **distributed-memory** systems where threads do not share one memory space and must explicitly send and receive data when necessary.

Reviewer: Technicality, but MPI is the standard rather than a library. The actual library could be OpenMPI, Intel MPI, etc.

#pragma omp parallel
{
  #pragma omp for
  for(...) {
    // perform some independent loop iterations
  }
}
```

Notice the new set of curly brackets which defines a new **scope**. At first this may seem strange, but scope is how OpenMP knows when to end the parallel region (or any other construct that applies to a *region* of code). You should already know about scope from your previous C++ knowledge so I won't explain it here. In C++ loops create their own new scope inside the loop delineated by curly braces, hence why we don't have to add any new braces when using the `parallel for`; it only applies to the duration of the for loop. But when using only the `parallel` construct, we have to explicitly tell the compiler where the parallel region starts and ends.

The `parallel` directive tells OpenMP that we want to parallelise the piece of code inside the following scope. What this really means is when execution reaches this parallel region, OpenMP will create a number of threads, allow any constructs inside the parallel region to use those threads, and then destroy those same threads. Strictly, the `parallel` construct *does nothing by itself* except create and destroy threads.

> The `for` construct in this example uses the threads spawned by `parallel` by splitting the iterations of the for loop into chunks and assigning different chunks to different threads. Default behaviour can change based on compiler and OpenMP runtime but usually the `for` construct will split the loop into as many chunks as their are threads and assign approximately the same number of iterations to each thread. So if there are 16 loop iterations to be performed at 4 threads, you can probably expect one thread to take iterations 0-4, another to take 5-8, and so on. Not all loops are this simple so we'll discuss later how we can optimise the `for` construct by specifying how the iterations are split amongst the threads.

Reviewer: ```suggestion
The `for` construct in this example uses the threads spawned by `parallel` by splitting the iterations of the for loop into chunks and assigning different chunks to different threads. Default behaviour can change based on the compiler and OpenMP runtime but usually the `for` construct will split the loop into as many chunks as there are threads and assign approximately the same number of iterations to each thread. So if there are 16 loop iterations to be performed at 4 threads, you can probably expect one thread to take iterations 0-4, another to take 5-8, and so on. Not all loops are this simple so we'll discuss later how we can optimise the `for` construct by specifying how the iterations are split amongst the threads.
```

### Example 1: Filling an array

Let's apply this `parallel for` to an example where we want to fill an array with values from a moderately expensive function like `sin`:

```cpp
#include <iostream>
#include <vector>
#include <cmath>
#include "timer.hpp"

using namespace std;

int main() {
  const int N = 100'000'000;
  vector<double> vec(N);

  Timer timer;

#pragma omp parallel for
  for(int i=0; i<vec.size(); ++i) {
    vec[i] = sin(M_PI*float(i)/N);
  }

  std::cout << timer.elapsed() << '\n';

  return 0;
}
```

You can find this example and all further examples in the examples repository TODO. The timer code in the example is relatively simple so I won't explain it in detail, it just starts a timer when it's created and shows the elapsed time when `timer.elapsed()` is called.

Let's build this *without* OpenMP to

1. show that the compiler will simply ignore the pragmas if we don't explicitly tell it to use OpenMP and
2. measure the time of the serial code.

> Building this with `g++ main.cpp timer.cpp` and running with `./a.out` prints out the time taken to fill this vector. It should take less than a second or so on your machine. Running on my laptop once, I got a time of $0.98$s. Adding the OpenMP flag for GCC, compiling with `g++ -fopenmp main.cpp timer.cpp` and again running with `./a.out` gives me a time of $0.14$. For me that's a speed up of around $700$% with one additional line of code! Given I have an 8-core processor, $800$% would be ideal but perfect scaling isn't usually possible in parallel computations, and perhaps with some optimisation we could squeeze out a little more performance. What performance increase do you see when running on your machine and does it match what you expect from the number of cores in your CPU?

Reviewer: This is clearer than the previous example, but "speed up" is still not technically defined, which it probably should be given that it's an important and standard measure.

Reviewer: You might want also to describe how they can check the number of cores on their CPU (or link to a source) as there's a good chance they won't know how many cores their CPU has!

I should point out here that this performance is when using the default compiler optimisation level `-O0`. If we switch to using `-O2` with `g++ -fopenmp -O2 main.cpp timer.cpp`, the serial performance drops to $0.74$s and the parallel performance to $0.10$, now giving a speedup of $740$%. While parallelisation can improve our overall performance greatly, single-core optimisations are still crucial to achieving the best performance possible from a code.

### Example 2: Summing an array (naively)

> Let's now sum the array from the previous example. I won't go into the maths but the sum of this actually gives us a numerical approximation of the integral of $sin(\pi x)$ from $0$ to $\pi$. If we sum the array, multiply it by $\pi$ and divide by $N$ we should get *exactly* $2$. The filling of the vector is the same so I'll only print here the code for the sum itself:

Reviewer: ```suggestion
Let's now sum the array from the previous example. I won't go into the maths but the sum of this actually gives us a numerical approximation of the integral of $sin(x)$ from $0$ to $\pi$. If we sum the array, multiply it by $\pi$ and divide by $N$ we should get *exactly* $2$. The filling of the vector is the same so I'll only print here the code for the sum itself:
```

```
  double sum = 0.;

  Timer timer;

#pragma omp parallel for
  for(int i=0; i<vec.size(); ++i) {
    sum += vec[i];
  }

  double elapsed = timer.elapsed();

  std::cout << "Time: " << elapsed << '\n';
  std::cout << "Result: " << M_PI*sum/N << '\n';
```

Compiling and running without OpenMP gives:

```
Time: 0.276116
Result: 2
```

This is exactly what we expect. If you enable OpenMP however, you'll probably get something like:

```
Time: 0.957646
Result: 0.14947
```

I recommend you take a break from reading/coding here and think about the questions, *why has the time increased with more threads* and *why is the answer wrong*? Don't worry if it's not obvious, this is a subtle and tricky bug and requires thinking about what each thread is doing as it accesses each variable inside the loop.

---

> This behaviour is the result of a **data race**. Consider even just two threads working in parallel inside the loop. Thread 1 reads data from `vec[i]`. The `for` construct has split the loop between the two threads so its value of `i` is totally unique to thread 1 and it is the only thread touching the data in `vec[i]`. It then needs to add this value into `sum`. It reads `sum`, adds `vec[i]` to it and writes it back into `sum`. This addition takes some time however, and thread 2 is performing the exact same operation but using a different value of `i`. When thread 1 reads `sum` and is currently spending time adding, thread 2 could easily have just finished an addition and has written a new value to `sum`. Thread 1 has no idea that thread 2 has done this and then unknowingly *overwrites* the updated value of `sum` with its own version. Not only does this mean the value of `sum` is *incorrect* but threads are not allowed to write to the same variable at the same time, so each threads must wait for the other to finish, adding more time to the overall calculation. This is why the overall time has increased, despite using multiple threads, and why the result is wrong!

Reviewer: ```suggestion
This behaviour is the result of a **data race**. Consider even just two threads working in parallel inside the loop. Thread 1 reads data from `vec[i]`. The `for` construct has split the loop between the two threads so its value of `i` is totally unique to thread 1 and it is the only thread touching the data in `vec[i]`. It then needs to add this value into `sum`. It reads `sum`, adds `vec[i]` to it and writes it back into `sum`. This addition takes some time however, and thread 2 is performing the exact same operation but using a different value of `i`. While thread 1 is occupied performing the addition, thread 2 could finish an addition operation of its own and write a new value to `sum`. Thread 1 has no idea that thread 2 has done this and then unknowingly *overwrites* the updated value of `sum` with its own version, thereby removing the contribution to the sum calculated by thread 2. Not only does this mean the value of `sum` is *incorrect* but threads are not allowed to write to the same variable at the same time, so each thread must wait for the other to finish, adding more time to the overall calculation. This is why the overall time has increased, despite using multiple threads, and why the result is wrong!
```
This wording is just a little clearer to me but totally up to you! 

In the longer term I think this would be great with some step by step diagrams to show how this works because it can be hard to visualise in your head if you're not very familiar with the way that CPUs work.

So how do we handle a data race? Each example of a data race is unique and must be handled in its own way, but the general solution is to limit threads to writing to variables no other threads will write to. We'll look later at a few different ways to tackle this particular problem but summing is part of a larger group of operations called **reductions**, where a large data structure (like a vector) is reduced to a smaller number of values, in this case just a single number. Thankfully, OpenMP provides a way to easily deal with these kinds of operations: reduction clauses.

### Example 3: Summing an array with `reduction`

> OpenMP allows constructs to be augmented with **clauses**, extra pieces of information we can add to its directives to further refine the parallelisation. In the above sum example, we can tell OpenMP we're trying to perform a reduction by adding the `reduction` clause to the `parallel for` construct along with the operator we want to reduce using and the variable that is holding the value of the reduction:

Reviewer: A bit more detail on this syntax would be good: what operators can they use? Can they define their own by passing in functions? Can they use lambda functions? 

Also whilst telling them that OpenMP can solve the problem using reductions, it would be valuable to tell them an idea of _how_ the problem is actually solved so they understand what is actually going on.

```
  double sum = 0.;

  Timer timer;

#pragma omp parallel for reduction(+:sum)
  for(int i=0; i<vec.size(); ++i) {
    sum += vec[i];
  }

  double elapsed = timer.elapsed();

  std::cout << "Time: " << elapsed << '\n';
  std::cout << "Result: " << M_PI*sum/N << '\n';
```

If you run the above example with OpenMP you should get something like:

```
Time: 0.0667325
Result: 2
```

### How to specify the number of threads in OpenMP?

### Fractal generation

[Fractals](https://en.wikipedia.org/wiki/Fractal) are shapes or patterns that in some sense contain infinite complexity. I won't go into their (very interesting) mathematical properties but plotting a fractal like the classic Mandelbrot set can produce some beautiful patterns:

![](img/mandelbrot.jpg)

> The above plot is made by turning each pixel with coordinates $(p,q)$ into a complex number $c = p + iq$ where $i^2 = -1$ is the imaginary unit. $c$ is used in the iterative equation $z_{n+1} = z_n^2 + c$ starting from $z_0 = 0$ and $z_n$ updated until either $|z_n| > 2$ or some maximum number of iterations is reached. It's not crucial to understand the maths behind this, 

Reviewer: the formatting of this has gone a bit haywire on the web build preview

## Numerically computing an integral

We can use common numerical techniques to turn an integral into a sum and compute than sum using a computer. Take calculating $\pi$ with the integral:

$\pi = 4\int_{0}^{1} \frac{dx}{1+x^2}$

Using the [midpoint rule](https://en.wikipedia.org/wiki/Numerical_integration#Quadrature_rules_based_on_interpolating_functions) we can turn this into a sum over $N$ pieces:

$\pi \approx \frac{1}{N}\sum_{i=1}{N} \frac{1}{1 + \left( \frac{i-1/2}{N} \right)^2}$

It's not too important to be able to derive this sum yourself if it's not your background, but you should be able to implement it in code. We could write the sum in C++ like:

```cpp
int N = 1024;
double sum = 0.0;
for (int i=1; i<N; ++i) {
  sum += 1.0 / ( 1.0 + pow((float(i) - 0.5)/N, 2.0) );
}
double pi = 4.0 * sum / N;
```

> $N=1024$ will run in a tiny amount of time on a modern CPU but if our integrand were more complex, if this were a 2D integral, if $N$ was extremely large, this sum could take long enough that it's worth parallelising. But how might we do that? The key here is noticing that the sum can be performed in any order, it fits our requirement of a parallel problem that pieces of the computation can be performed independently.

Reviewer: ```suggestion
$N=1024$ will run in a tiny amount of time on a modern CPU but if our integrand were more complex, if this were a 2D integral, if $N$ were extremely large, this sum could take long enough that it's worth parallelising. But how might we do that? The key here is noticing that the sum can be performed in any order, it fits our requirement of a parallel problem that pieces of the computation can be performed independently.
```
